lib/Dialect/AIEVec/Transforms/ConvertVectorToAIEVec.cpp

```diff
@@ -0,0 +1,244 @@
+//===-ConvertVectorToAIEVec.cpp - Lower Vector to AIE vector ----*- C++ -*-===//
+//
+// This file is licensed under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+// (c) Copyright 2022 Xilinx Inc.
+//
+//===----------------------------------------------------------------------===//
+// This is the implementation of the lowering pass from standard Vector
+// dialect to AIEVec, compatible with the AIE vector architecture.
+//===----------------------------------------------------------------------===//
+
+#include <algorithm>
+
+#include "aie/Dialect/AIEVec/AIEVecUtils.h"
+#include "aie/Dialect/AIEVec/IR/AIEVecOps.h"
+#include "aie/Dialect/AIEVec/Transforms/IntervalReuse.h"
+#include "aie/Dialect/AIEVec/Transforms/Passes.h"
+#include "mlir/Conversion/AffineToStandard/AffineToStandard.h"
+#include "mlir/Dialect/Affine/Analysis/LoopAnalysis.h"
+#include "mlir/Dialect/Affine/IR/AffineOps.h"
+#include "mlir/Dialect/Func/IR/FuncOps.h"
+#include "mlir/Dialect/MemRef/IR/MemRef.h"
+#include "mlir/Dialect/SCF/IR/SCF.h"
+#include "mlir/Dialect/Vector/Transforms/VectorTransforms.h"
+#include "mlir/IR/TypeUtilities.h"
+#include "mlir/Pass/PassManager.h"
+#include "mlir/Transforms/DialectConversion.h"
+#include "mlir/Transforms/Passes.h"
+#include "llvm/ADT/SmallSet.h"
+
+namespace xilinx::aievec {
+#define GEN_PASS_DEF_CONVERTVECTORTOAIEVEC
+#include "aie/Dialect/AIEVec/Transforms/Passes.h.inc"
+} // namespace xilinx::aievec
+
+using namespace mlir;
+using namespace arith;
+using namespace vector;
+using namespace xilinx;
+using namespace xilinx::aievec;
+
+#define DEBUG_TYPE "aievec-lowering"
+
+//===----------------------------------------------------------------------===//
+// Analyses
+//===----------------------------------------------------------------------===//
+
+struct UPDOpEffectiveAccessSizeAnalysis {
+  UPDOpEffectiveAccessSizeAnalysis(aievec::UPDOp updOp) {
+    auto vecType = cast<VectorType>(updOp.getResult().getType());
+    unsigned sizeInBits =
+        cast<ShapedType>(vecType).getSizeInBits() - updOp.getOffset();
+    for (Operation *user : updOp->getUsers()) {
+      auto userUpdOp = dyn_cast<xilinx::aievec::UPDOp>(user);
+      if (userUpdOp)
+        sizeInBits -= userUpdOp.getOffset();
+    }
```

Comment:
Not quite understand line 55-59 for the upd which is the user of a upd , could you please add a simple case in the comment?

Reply:
This code is a bit hackish, and it's supposed to go eventually. It's just a little trick to evaluate whether two or more UPD ops are linked, but it's just easier to generate them when lowering transfer_read. I will add an explanation, but I don't think it's worth thinking too hard about this.